core/blockchain.go

```diff
@@ -1723,6 +1761,10 @@ func (bc *BlockChain) insertChain(chain types.Blocks, setHead bool) (int, error)
 				return it.index, err
 			}
 			stats.processed++
+			if bc.logger != nil {
+				bc.logger.OnBlockStart(block, bc.GetTd(block.ParentHash(), block.NumberU64()-1), bc.CurrentFinalBlock(), bc.CurrentSafeBlock(), bc.chainConfig)
+				bc.logger.OnBlockEnd(nil)
+			}
```

Comment:
The reason why we added it is Firehose experienced some cases that blocks were missing (i.e. they didn't get the block events) on polygon (this goes back to before the extended tracer PR). So they'd like to have this. I haven't been able to confirm or refute this.

We just discussed that in case of setHead known blocks would be announced again. I think this is ok. Just I'd like to add a parameter to OnBlockStart to indicate this is a known block.

Reply:
I decided to add a flag `skip` to "OnBlockStart" to indicate the block is known and will not be processed, i.e. consumers should not expect further Transaction/call/etc. events for this block.

I am keeping this event, so it's possible to determine the progression of the chain as geth sees it. E.g. the catch-up phase after a crash.

One more reason is I'm seeing in the code that a known block can trigger a reorg:

https://github.com/ethereum/go-ethereum/blob/8a76a814a2b9e5b4c1a4c6de44cd702536104507/core/blockchain.go#L1333-L1339

In this case it's important the consumer can figure out we went on a sidechain.